include/nbl/video/utilities/CComputeBlit.h

```diff
@@ -0,0 +1,900 @@
+				return nullptr;
+		}
+
+		result->m_coverageAdjustmentPipelineLayout = result->m_device->createPipelineLayout(&pcRange, &pcRange + 1ull, core::smart_refctd_ptr(result->m_blitDSLayout[EBT_COVERAGE_ADJUSTMENT]));
+		if (!result->m_coverageAdjustmentPipelineLayout)
+			return nullptr;
+
+		return result;
+	}
+
+	inline void setAvailableSharedMemory(const uint32_t smemSize)
+	{
+		if (smemSize == ~0u)
+			m_availableSharedMemory = m_device->getPhysicalDevice()->getProperties().limits.maxComputeSharedMemorySize;
+		else
+			m_availableSharedMemory = core::min(core::roundUp(smemSize, static_cast<uint32_t>(sizeof(float) * 64)), m_device->getPhysicalDevice()->getLimits().maxComputeSharedMemorySize);
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> getDefaultBlitDescriptorSetLayout(const asset::IBlitUtilities::E_ALPHA_SEMANTIC alphaSemantic) const
+	{
+		if (alphaSemantic == asset::IBlitUtilities::EAS_REFERENCE_OR_COVERAGE)
+			return m_blitDSLayout[EBT_COVERAGE_ADJUSTMENT];
+		else
+			return m_blitDSLayout[EBT_REGULAR];
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> getDefaultKernelWeightsDescriptorSetLayout() const
+	{
+		return m_kernelWeightsDSLayout;
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUPipelineLayout> getDefaultBlitPipelineLayout(const asset::IBlitUtilities::E_ALPHA_SEMANTIC alphaSemantic) const
+	{
+		if (alphaSemantic == asset::IBlitUtilities::EAS_REFERENCE_OR_COVERAGE)
+			return m_blitPipelineLayout[EBT_COVERAGE_ADJUSTMENT];
+		else
+			return m_blitPipelineLayout[EBT_REGULAR];
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUPipelineLayout> getDefaultCoverageAdjustmentPipelineLayout() const
+	{
+		return m_coverageAdjustmentPipelineLayout;
+	}
+
+	// @param `alphaBinCount` is only required to size the histogram present in the default nbl_glsl_blit_AlphaStatistics_t in default_compute_common.comp
+	core::smart_refctd_ptr<video::IGPUSpecializedShader> createAlphaTestSpecializedShader(const asset::IImage::E_TYPE inImageType, const uint32_t alphaBinCount = DefaultAlphaBinCount);
+
+	core::smart_refctd_ptr<video::IGPUComputePipeline> getAlphaTestPipeline(const uint32_t alphaBinCount, const asset::IImage::E_TYPE imageType)
+	{
+		const auto workgroupDims = getDefaultWorkgroupDims(imageType);
+		const auto paddedAlphaBinCount = getPaddedAlphaBinCount(workgroupDims, alphaBinCount);
+
+		assert(paddedAlphaBinCount >= MinAlphaBinCount);
+		const auto pipelineIndex = (paddedAlphaBinCount / MinAlphaBinCount) - 1;
+
+		if (m_alphaTestPipelines[pipelineIndex][imageType])
+			return m_alphaTestPipelines[pipelineIndex][imageType];
+
+		auto specShader = createAlphaTestSpecializedShader(imageType, paddedAlphaBinCount);
+		m_alphaTestPipelines[pipelineIndex][imageType] = m_device->createComputePipeline(nullptr, core::smart_refctd_ptr(m_blitPipelineLayout[EBT_COVERAGE_ADJUSTMENT]), std::move(specShader));
+
+		return m_alphaTestPipelines[pipelineIndex][imageType];
+	}
```

Comment:
are you using the `histogram` in the alpha test shader?

you know that you can always use a DS+Pipeline layout that has bindings which are not declared (and ergo not used) in the shader itself?

Reply:
> are you using the histogram in the alpha test shader?

No, but `nbl_glsl_blit_AlphaStatistics_t` requires `_NBL_GLSL_BLIT_ALPHA_BIN_COUNT_` to have the correct memory layout in the passed SSBO --this comes into play for multiple layers.

Reply:
And the fact that `nbl_glsl_blit_AlphaStatistics_t` is common for all three steps.

Reply:
you can lie about the size of `_NBL_GLSL_BLIT_ALPHA_BIN_COUNT_` if you're not using `nbl_glsl_blit_AlphaStatistics_t` anyway :smart:

Reply:
It will be used implicitly.

The interface block (for the Alpha Test shader pass) for the scratch SSBO looks something like this:
```GLSL
layout(set = n, binding = m) restrict coherent buffer AlphaTestPassedCounterBuffer
{
	nbl_glsl_blit_AlphaStatistics_t data[];
} alphaTestPassedCounterBuffer;
```
and `nbl_glsl_blit_AlphaStatistics_t` is defined as:
```GLSL
struct nbl_glsl_blit_AlphaStatistics_t
{
	uint passedPixelCount;
	uint histogram[_NBL_GLSL_BLIT_ALPHA_BIN_COUNT_];
};
```
Where the size of the array `nbl_glsl_blit_AlphaStatistics_t data[];` depends on the number of layers an image has.
If you use any other value of `_NBL_GLSL_BLIT_ALPHA_BIN_COUNT_` then `sizeof(nbl_glsl_blit_AlphaStatistics_t)` will differ which in turn will mess up the memory layout, and consequently, the next layer's `passedPixelCount` might be read from some other layer's `histogram` member, and other strange things like that.

All of this is given we bound the same scratch SSBO for both Alpha Test and Normalization passes.

I think we already had this discussion before. 🤔

Reply:
ah ok, I forgot the memory layout of that SSSBO, you're right